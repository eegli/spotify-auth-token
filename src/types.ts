type OptionalConfig = {
  port: number;
  scopes: string;
  fileName: string;
  outDir: string;
  noEmit: boolean;
};

type RequiredConfig = {
  clientId: string;
  clientSecret: string;
};

export type AuthFunction<T extends UserConfig = UserConfig> = (
  config?: T
) => Promise<SpotifyTokenResponse>;

export type UserConfig = RequiredConfig & Partial<OptionalConfig>;
export type AppConfig = RequiredConfig & OptionalConfig;

export type SpotifyTokenResponse = {
  access_token: string;
  token_type: 'Bearer';
  expires_in: number;
  refresh_token: string;
  scope: string;
  date_obtained: string;
};
