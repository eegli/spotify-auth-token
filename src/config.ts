import { parserFactory } from '@eegli/tinyparse';
import { AppConfig } from './types';

export const defaultConfig: AppConfig = {
  clientId: '',
  clientSecret: '',
  port: 3000,
  outDir: '',
  fileName: 'spotify-token',
  scopes: 'user-read-email',
};

export const configParser = parserFactory(defaultConfig, {
  required: ['clientId', 'clientSecret'],
  shortFlags: {
    '-ci': 'clientId',
    '-cs': 'clientSecret',
    '-p': 'port',
    '-o': 'outDir',
    '-f': 'fileName',
    '-s': 'scopes',
  },
});
