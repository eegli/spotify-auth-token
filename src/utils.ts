import { existsSync, promises as fs } from 'fs';
import { join } from 'path';

export async function write(
  path: string,
  fileName: string,
  data: unknown
): Promise<string> {
  path = join(process.cwd(), path);
  if (!existsSync(path)) {
    await fs.mkdir(path, { recursive: true });
  }
  if (fileName.endsWith('.json')) {
    fileName = fileName.slice(0, -5);
  }
  path = join(path, fileName + '.json');
  await fs.writeFile(path, JSON.stringify(data, null, 2));
  return path;
}

export function id(): string {
  return Math.random().toString(36).slice(2);
}

export function goodbye(message: string): never {
  console.error('\x1b[31m', `Error: ${message}. Goodbye`, '\x1b[0m');
  process.exit(1);
}
