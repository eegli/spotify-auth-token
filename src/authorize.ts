import { ValidationError } from '@eegli/tinyparse';
import { configParser } from './config';
import { getLocalhostUrl, request } from './request';
import type { SpotifyTokenResponse, UserConfig } from './types';
import { goodbye, id, write } from './utils';

export async function authorize(
  userConfig: UserConfig
): Promise<SpotifyTokenResponse | undefined> {
  try {
    const config = userConfig
      ? await configParser(userConfig)
      : await configParser(process.argv.slice(2));

    const redirectUri = `http://localhost:${config.port}`;
    const state = id();

    const spotifyUrl =
      'https://accounts.spotify.com/authorize?' +
      new URLSearchParams({
        response_type: 'code',
        show_dialog: 'true',
        state: encodeURIComponent(state),
        client_id: encodeURIComponent(config.clientId),
        redirect_uri: redirectUri,
        scope: encodeURIComponent(config.scopes),
      }).toString();

    console.info('Please click the link to login to Spotify in the browser\n');
    console.info(spotifyUrl + '\n');

    const authUrl = await getLocalhostUrl(config.port);
    const params = new URLSearchParams(authUrl);
    const receivedCode = params.get('code');
    const receivedState = params.get('state');

    if (receivedState !== state) {
      goodbye('Received and original state do not match');
    }

    if (!receivedCode) {
      goodbye('No code received');
    }

    console.info('Login successfull! Cleaning up...\n');

    const tokenRequestBody = new URLSearchParams({
      grant_type: 'authorization_code',
      code: receivedCode,
      redirect_uri: redirectUri,
    });

    const token: SpotifyTokenResponse = await request(
      {
        hostname: 'accounts.spotify.com',
        path: '/api/token',
        method: 'POST',
        headers: {
          'Content-type': 'application/x-www-form-urlencoded',
          Authorization:
            'Basic ' +
            Buffer.from(config.clientId + ':' + config.clientSecret).toString(
              'base64'
            ),
        },
      },
      tokenRequestBody.toString()
    );

    token.date_obtained = new Date().toUTCString();

    if (config.noEmit) {
      console.info('Success!');
      return token;
    } else {
      const outDir = await write(config.outDir, config.fileName, token);
      console.info(`Success! Saved Spotify access token to "${outDir}"`);
    }
  } catch (e) {
    if (e instanceof ValidationError) {
      goodbye(e.message);
    }
    goodbye('Something went wrong: ' + e);
  }
}
