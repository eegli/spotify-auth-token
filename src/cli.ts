import { authorize } from './authorize';
import { exit } from './utils';

export const cli = async () => {
  if (process.argv.length < 3) {
    exit(
      `
spotify-auth-token

- A command line utility for generating Spotify refresh tokens.

For docs & help, visit https://github.com/eegli/spotify-auth-token.
    `,
      'yellow'
    );
  }

  await authorize();
};
