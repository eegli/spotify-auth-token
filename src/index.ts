import { authorize } from './authorize';

export type { UserConfig } from './types';
export default authorize;
