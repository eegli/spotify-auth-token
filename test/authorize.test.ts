import auth from '../src/index';
import * as request from '../src/request';
import * as utils from '../src/utils';

const mockRequest = request as jest.Mocked<typeof request>;

// Math.random().toString(36).slice(2) now returns "ou8n1fu8n1"
jest.spyOn(Math, 'random').mockReturnValue(0.69);

const testState = 'ou8n1fu8n1';

mockRequest.getLocalhostUrl.mockResolvedValue(
  `?code=AQDKHwNyRapw&state=${testState}`
);

mockRequest.request.mockResolvedValue({
  access_token: 'BQC2fMYf9',
  token_type: 'Bearer',
  expires_in: 3600,
  refresh_token: 'OFuVRQAzVzq0Wy_Py_W4KiMr8H0',
  scope: 'user-library-read user-top-read',
});

const consoleSpy = jest
  .spyOn(global.console, 'info')
  .mockImplementation(jest.fn());

const writeSpy = jest
  .spyOn(utils, 'write')
  .mockImplementation((path) => Promise.resolve(path));

afterEach(() => {
  jest.clearAllMocks();
});

describe('Authorize with params', () => {
  it('fails with missing clientId', async () => {
    // @ts-expect-error test input
    await expect(auth({ clientId: 'id' })).rejects.toThrow();
  });

  it('fails with missing clientSecret', async () => {
    // @ts-expect-error test input
    await expect(auth({ clientSecret: 'secret' })).rejects.toThrow();
  });

  it("fails if states don't match", async () => {
    mockRequest.getLocalhostUrl.mockResolvedValueOnce(
      `?code=AQDKHwNyRapw&state=${testState}XXX`
    );
    await expect(
      auth({ clientId: 'cid', clientSecret: 'cs' })
    ).rejects.toThrow();
  });

  it('fails if no code is received', async () => {
    mockRequest.getLocalhostUrl.mockResolvedValueOnce(`?state=${testState}`);
    await expect(
      auth({ clientId: 'cid', clientSecret: 'cs' })
    ).rejects.toThrow();
  });

  it('works with emit mode', async () => {
    const config = {
      clientId: 'cid',
      clientSecret: 'cs',
      port: 1000,
      outDir: 'out/token/',
      fileName: 'mytoken',
      scopes: 'user-do-nothing user-balabla',
    };
    const result = await auth(config);
    expect(result).toBeDefined();
    expect(consoleSpy.mock.calls[1][0]).toMatchSnapshot('auth url');
    expect(mockRequest.getLocalhostUrl).toHaveBeenCalledWith(config.port);
    expect(mockRequest.request.mock.calls[0][0]).toMatchSnapshot(
      'spotify request'
    );
    expect(writeSpy.mock.calls[0]).toMatchSnapshot('write data');
    expect(consoleSpy.mock.calls[3][0]).toMatch(new RegExp(config.outDir));
  });

  it('works with no emit mode', async () => {
    const result = await auth({
      clientId: 'cid',
      clientSecret: 'cs',
      noEmit: true,
    });
    expect(result).toBeDefined();
    expect(writeSpy).not.toHaveBeenCalled();
  });
});

describe('Authorize with process.argv', () => {
  it('fails with missing clientId', async () => {
    process.argv = ['', ''];
    // @ts-expect-error - get args from process.argv
    await expect(auth()).rejects.toThrow();
  });
  [
    ['', '', '--clientId', '111x', '--clientSecret', '111x'],
    ['', '', '--clientId', '333x', '--clientSecret', '333x', '--port', '4000'],
  ].forEach(async (args) => {
    process.argv = args;
    // @ts-expect-error - get args from process.argv
    await expect(auth()).resolves.not.toThrow();
  });
});
