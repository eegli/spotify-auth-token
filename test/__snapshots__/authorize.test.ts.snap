// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorize with params works with config 0: auth url 1`] = `
"https://accounts.spotify.com/authorize?response_type=code&show_dialog=true&state=ou8n1fu8n1&client_id=cid&redirect_uri=http%3A%2F%2Flocalhost%3A1000&scope=user-do-nothing
"
`;

exports[`Authorize with params works with config 0: spotify request 1`] = `
Object {
  "headers": Object {
    "Authorization": "Basic Y2lkOmNz",
    "Content-type": "application/x-www-form-urlencoded",
  },
  "hostname": "accounts.spotify.com",
  "method": "POST",
  "path": "/api/token",
}
`;

exports[`Authorize with params works with config 0: write data 1`] = `
Array [
  "out/token/",
  "mytoken",
  Object {
    "access_token": "BQC2fMYf9",
    "date_obtained": "Sat, 20 Apr 1996 22:00:00 GMT",
    "expires_in": 3600,
    "refresh_token": "OFuVRQAzVzq0Wy_Py_W4KiMr8H0",
    "scope": "user-library-read",
    "token_type": "Bearer",
  },
]
`;

exports[`Authorize with params works with config 1: auth url 1`] = `
"https://accounts.spotify.com/authorize?response_type=code&show_dialog=true&state=ou8n1fu8n1&client_id=cid&redirect_uri=http%3A%2F%2Flocalhost%3A59&scope=user-read-a-book
"
`;

exports[`Authorize with params works with config 1: spotify request 1`] = `
Object {
  "headers": Object {
    "Authorization": "Basic Y2lkOmNz",
    "Content-type": "application/x-www-form-urlencoded",
  },
  "hostname": "accounts.spotify.com",
  "method": "POST",
  "path": "/api/token",
}
`;

exports[`Authorize with params works with config 1: write data 1`] = `
Array [
  "/out/token/",
  "token",
  Object {
    "access_token": "BQC2fMYf9",
    "date_obtained": "Sat, 20 Apr 1996 22:00:00 GMT",
    "expires_in": 3600,
    "refresh_token": "OFuVRQAzVzq0Wy_Py_W4KiMr8H0",
    "scope": "user-library-read",
    "token_type": "Bearer",
  },
]
`;

exports[`Authorize with params works with config 2: auth url 1`] = `
"https://accounts.spotify.com/authorize?response_type=code&show_dialog=true&state=ou8n1fu8n1&client_id=cid&redirect_uri=http%3A%2F%2Flocalhost%3A69&scope=scropes2
"
`;

exports[`Authorize with params works with config 2: spotify request 1`] = `
Object {
  "headers": Object {
    "Authorization": "Basic Y2lkOmNz",
    "Content-type": "application/x-www-form-urlencoded",
  },
  "hostname": "accounts.spotify.com",
  "method": "POST",
  "path": "/api/token",
}
`;

exports[`Authorize with params works with config 2: write data 1`] = `
Array [
  "/out/",
  "spotify-token.json",
  Object {
    "access_token": "BQC2fMYf9",
    "date_obtained": "Sat, 20 Apr 1996 22:00:00 GMT",
    "expires_in": 3600,
    "refresh_token": "OFuVRQAzVzq0Wy_Py_W4KiMr8H0",
    "scope": "user-library-read",
    "token_type": "Bearer",
  },
]
`;
